### PR TITLE
Stablecoin investment tests

### PR DESCRIPTION
**Pull request by juztamau5**
## Description

This PR adds Yearn and Curve as dependencies, implements stablecoin investing in CFIHSC, and adds test cases that correspond to, and expand upon, LP NFT tests.

I had to guess on some numbers, so they might need adjusting.

Note that USDT is not ERC-20 compliant, and has special handling in Curve. We might also need special handling in our code (for now the special handling is patched out in Curve). I marked such places with TODOs.

## How has this been tested?

Developed using CI, so runs should be green

### CODE REVIEW

File: contracts/src/cfolio/CFolioItemHandlerSC.sol [hunk 1]
```diff
@@ -216,15 +233,44 @@ contract CFolioItemHandlerSC is ICFolioItemHandler, Context {
     // Validate parameters, no unmasking required, must be SFT
     address cFolio = sftHolder.tokenIdToAddress(sftTokenId);
     require(cFolio != address(0), 'Invalid sftTokenId');
-
-    require(amounts.length == 0, 'CFIH: not impl');
+    if (amounts.length > 0) {
+      require(amounts.length == 4, 'Need DAI/USDC/USDT/TUSD');
```
Comment (peak3d) [inline contracts/src/cfolio/CFolioItemHandlerSC.sol]: require(amounts.length == 0 || amounts.length == 4, 'Need DAI/USDC/USDT/TUSD');

File: contracts/src/cfolio/CFolioItemHandlerSC.sol [hunk 2]
```diff
@@ -250,10 +296,58 @@ contract CFolioItemHandlerSC is ICFolioItemHandler, Context {
...
     // Validate parameters
-    require(amounts.length == 0, 'CFIH: Not impl');
-    _verifyAssetAccess(baseTokenId, tokenId);
+    require(amounts.length == 4, 'Need DAI/USDC/USDT/TUSD');
+    (address baseCFolio, address itemCFolio) =
+      _verifyAssetAccess(baseTokenId, tokenId);
+
+    // Transfer stablecoins to this contract
+    uint256[4] memory tokenAmounts;
+    for (uint256 i = 0; i < 4; ++i) {
```
Comment (peak3d) [inline contracts/src/cfolio/CFolioItemHandlerSC.sol]: I suspect that normal user only deposits one coin, from reading logs of curve, it cost quite some gas to call transfer for 0 value, IMO we should condition this here (and dont call all the stuff if no value given))

File: contracts/src/cfolio/CFolioItemHandlerSC.sol [hunk 3]
```diff
@@ -250,10 +296,58 @@ contract CFolioItemHandlerSC is ICFolioItemHandler, Context {
...
+      IERC20(underlyingCoin).transferFrom(
+        _msgSender(),
+        address(this),
+        amounts[i]
+      );
+
+      tokenAmounts[i] = IERC20(underlyingCoin).balanceOf(address(this)); // Handle deflationary coins
+
+      // Approve funds
+      IERC20(underlyingCoin).approve(address(curveYDeposit), tokenAmounts[i]);
```
Comment (peak3d) [inline contracts/src/cfolio/CFolioItemHandlerSC.sol]: Is this really required? I would assume that this is already done internally?
Edit: Nvm

File: contracts/src/cfolio/CFolioItemHandlerSC.sol [hunk 4]
```diff
@@ -269,9 +363,53 @@ contract CFolioItemHandlerSC is ICFolioItemHandler, Context {
...
+
+    // Record assets in Farm contract. They don't earn rewards.
+    // addAsset must only be called from Investment CFolios
+    cfolioFarm.removeAssets(itemCFolio, balanceBefore.sub(balanceAfter));
+
+    if (baseTokenId != uint256(-1))
+      _updateRewards(baseCFolio, sftEvaluator.rewardRate(baseTokenId));
+
+    // Transfer stablecoins back to the sender
+    for (uint256 i = 0; i < 4; ++i) {
```
Comment (peak3d) [inline contracts/src/cfolio/CFolioItemHandlerSC.sol]: If we only allow one coin removal, then there is no loop required, correct? coinIndex should already point to the correct asset

**Review by peak3d — CHANGES_REQUESTED** [review]
Great work! There are some minor things to clearify, and I haven't looked that deep into the test system.
1.) we'll need the 2 new addresses also in generated_addresses.json in order to predefine them for the mainnet deployment (so same concept you implemented for OSProxyRegistry for example, or the Uniswap things).
2.) discussion point would be for both deposit and withdraw if we want to allow it for already existing YToken. maybe there are ppl. which have already something in curve, and want to earn wows without the expensive withdraw deposit procedure.
3.) We finally have to fixate the team invest things,  it is communicated to the crowd, so we have to implement it before we can pass it to audit.

**Comment by juztamau5** [comment]
Sorry for the delay, haven't been able to make it to a computer for four days. I'm starting to address the comments now. Here's some responses:

> 1.) we'll need the 2 new addresses also in generated_addresses.json in order to predefine them for the mainnet deployment (so same concept you implemented for OSProxyRegistry for example, or the Uniswap things).

I intended for this PR to include predefining the addresses, I'll look closer and see if that's actually the case.

> 2.) discussion point would be for both deposit and withdraw if we want to allow it for already existing YToken. maybe there are ppl. which have already something in curve, and want to earn wows without the expensive withdraw deposit procedure.

This is a good idea, but one problem is that we want to take a small percentage of the profit. If you invest some dai and get back more dai, then you can calculate a fee, because there was a profit in the deposited token. If you invest YToken and get back same amount of YToken, you can't take a fee, because YTokens increase in value, not in amount, so you'd be taking some of the principal.

I realized that we don't currently take a profit fee. I'm considering how this would be done, but I suggest we merge the PR and then follow up with this feature.

File: contracts/src/cfolio/CFolioItemHandlerSC.sol (hunk 2 above)
Comment (juztamau5) [reply contracts/src/cfolio/CFolioItemHandlerSC.sol]: Added a check. Surprisingly, in the test fixture, it costs less gas to deposit four even amounts than a single non-zero amount because less swapping is required.

File: contracts/src/cfolio/CFolioItemHandlerSC.sol (hunk 1 above)
Comment (juztamau5) [reply contracts/src/cfolio/CFolioItemHandlerSC.sol]: Fixed, thanks

File: contracts/src/cfolio/CFolioItemHandlerSC.sol (hunk 4 above)
Comment (juztamau5) [reply contracts/src/cfolio/CFolioItemHandlerSC.sol]: Thanks, fixed

**Comment by juztamau5** [comment]
@peak3d pushed updates for your comments. Also added a commit to seed liquidity, as tests were unable to fully convert everything to DAI when the pool is empty. Also, I confirmed that unless the `IGNORE_ADDRESSES` is set, deployment will read addresses from `addresses.json`.

File: contracts/src/cfolio/CFolioItemHandlerSC.sol [hunk 5]
```diff
@@ -216,15 +234,45 @@ contract CFolioItemHandlerSC is ICFolioItemHandler, Context {
...
+      // Transfer stablecoins to this contract
+      for (uint256 i = 0; i < 4; ++i) {
+        address underlyingCoin = curveYDeposit.underlying_coins(int128(i));
+
+        // TODO: Special handling for USDT?
+        // TODO: Safe transfer?
+        IERC20(underlyingCoin).transferFrom(payer, address(this), amounts[i]);
+
+        uint256 newBalance = IERC20(underlyingCoin).balanceOf(address(this));
+        IERC20(underlyingCoin).approve(address(curveYDeposit), newBalance);
```
Comment (peak3d) [inline contracts/src/cfolio/CFolioItemHandlerSC.sol]: Please make unlimited approvals once during contract initialization

File: contracts/src/cfolio/CFolioItemHandlerSC.sol [hunk 6]
```diff
@@ -250,10 +298,70 @@ contract CFolioItemHandlerSC is ICFolioItemHandler, Context {
...
+      IERC20(underlyingCoin).transferFrom(
+        _msgSender(),
+        address(this),
+        amounts[i]
+      );
+
+      uint256 tokenAmount = IERC20(underlyingCoin).balanceOf(address(this)); // Handle deflationary coins
+
+      // Approve funds
+      IERC20(underlyingCoin).approve(address(curveYDeposit), tokenAmount);
```
Comment (peak3d) [inline contracts/src/cfolio/CFolioItemHandlerSC.sol]: Please make unlimited approvals once during contract initialization

File: contracts/src/cfolio/CFolioItemHandlerSC.sol [hunk 7]
```diff
@@ -269,9 +377,51 @@ contract CFolioItemHandlerSC is ICFolioItemHandler, Context {
...
+        amount = amounts[i];
+      }
+    }
+
+    // Validate parameters
+    require(coinIndex != -1, 'All amounts are zero');
+
+    // Call to external contract
+    uint256 balanceBefore = curveYToken.balanceOf(address(this));
+    curveYToken.approve(address(curveYDeposit), amount);
```
Comment (peak3d) [inline contracts/src/cfolio/CFolioItemHandlerSC.sol]: Please make unlimited approvals once during contract initialization

**Comment by juztamau5** [comment]
TODO (listing so I can keep track):

* Approvals in constructor, proxy initialization
* Accept 5 amounts (DAI/USDC/USDT/TUSD/yCRV)
* getAmounts() for SFT minter
* min amount for withdrawal
* yCRV gauge (stretch goal)

**Comment by peak3d** [comment]
> TODO (listing so I can keep track):
> 
> * Approvals in constructor, proxy initialization
> * Accept 5 amounts (DAI/USDC/USDT/TUSD/yCRV)
> * getAmounts() for SFT minter
> * min amount for withdrawal
> * yCRV gauge (stretch goal)

Pls. remove function calcWithdrawOneCoin(uint256, uint128)

**Comment by juztamau5** [comment]
@peak3d I was able to address all the comments I could track down. This should be ready for UI testing, unless you have more comments.

**Comment by juztamau5** [comment]
Actually still one TODO left: https://github.com/wolvesofwallstreet/wolves.finance/pull/117/files#diff-d753fb665839f96d7328712456b8f9c2484092835668b67e85fc9b3f874f05bfR143
